UefiPayloadPkg/Library/FdtParserLib/FdtParserLib.c

```diff
@@ -0,0 +1,627 @@
+          if (TempLen > 0) {
+            Data32 = (UINT32 *)(PropertyPtr->Data);
+            DEBUG ((DEBUG_INFO, "\n         Property(00000000)  address_width"));
+            DEBUG ((DEBUG_INFO, "  %X", Fdt32ToCpu (*Data32)));
+            SizeOfMemorySpace = (UINT8)Fdt32ToCpu (*Data32);
+ #if defined (MDE_CPU_IA32) || defined (MDE_CPU_X64)
+            BuildCpuHob (SizeOfMemorySpace, 16);
+ #else
+            BuildCpuHob (SizeOfMemorySpace, 0);
+ #endif
+          }
+        }
+      }
+    }
+    // Optional
+    else if (AsciiStrnCmp (NodePtr->Name, "pci-rb", AsciiStrLen ("pci-rb")) == 0) {
+      DEBUG ((DEBUG_INFO, "  Found pci-rb node (%08X)", Node));
+
+      HobDataSize = sizeof (UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES) + RootBridgeCount *sizeof (UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGE);
+      //
+      // Create PCI Root Bridge Info Hob.
+      //
+      if (PciRootBridgeInfo == NULL) {
+        PciRootBridgeInfo = BuildGuidHob (&gUniversalPayloadPciRootBridgeInfoGuid, HobDataSize);
+        ASSERT (PciRootBridgeInfo != NULL);
+        if (PciRootBridgeInfo == NULL) {
+          break;
+        }
+
+        ZeroMem (PciRootBridgeInfo, HobDataSize);
+        PciRootBridgeInfo->Header.Length    = sizeof (UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES);
+        PciRootBridgeInfo->Header.Revision  = UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES_REVISION;
+        PciRootBridgeInfo->Count            = RootBridgeCount;
+        PciRootBridgeInfo->ResourceAssigned = FALSE;
+      }
+
+      for (Property = FdtFirstPropertyOffset (Fdt, Node); Property >= 0; Property = FdtNextPropertyOffset (Fdt, Property)) {
+        PropertyPtr = FdtGetPropertyByOffset (Fdt, Property, &TempLen);
+        TempStr     = FdtGetString (Fdt, Fdt32ToCpu (PropertyPtr->NameOffset), NULL);
+
+        if (AsciiStrCmp (TempStr, "ranges") == 0) {
+          DEBUG ((DEBUG_INFO, "  Found ranges Property TempLen (%08X)\n", TempLen));
+
+          PciRootBridgeInfo->RootBridge[index].AllocationAttributes = EFI_PCI_HOST_BRIDGE_COMBINE_MEM_PMEM | EFI_PCI_HOST_BRIDGE_MEM64_DECODE;
+          PciRootBridgeInfo->RootBridge[index].Supports             = ROOT_BRIDGE_SUPPORTS_DEFAULT;
+          PciRootBridgeInfo->RootBridge[index].PMemAbove4G.Base     = PcdGet64 (PcdPciReservedPMemAbove4GBBase);
+          PciRootBridgeInfo->RootBridge[index].PMemAbove4G.Limit    = PcdGet64 (PcdPciReservedPMemAbove4GBLimit);
+          PciRootBridgeInfo->RootBridge[index].PMem.Base            = PcdGet32 (PcdPciReservedPMemBase);
+          PciRootBridgeInfo->RootBridge[index].PMem.Limit           = PcdGet32 (PcdPciReservedPMemLimit);
+          PciRootBridgeInfo->RootBridge[index].UID                  = index;
+          PciRootBridgeInfo->RootBridge[index].HID                  = EISA_PNP_ID (0x0A03);
+
+          Data32                                                = (UINT32 *)(PropertyPtr->Data);
+          PciRootBridgeInfo->RootBridge[index].Mem.Base         = Fdt32ToCpu (*(Data32 + 2));

```

Comment:
Is this a hard assumption that first entry is always mem base? It should be based on resource type value.

Reply:
we have to check ss bits to see if the range data is I/O, MMIO32 or MMIO64, following pci_bus binding document (https://www.openfirmware.info/data/docs/bus.pci.pdf):

Encoding of type code "ss":
00 denotes Configuration Space, in which case:
...
01 denotes I/O Space, in which case:
...
10 denotes 32-bit-address Memory Space, in which case:
...
11 denotes 64-bit-address Memory Space, in which case:
...